grain/_src/python/dataset/transformations/prefetch_test.py

```diff
@@ -1148,6 +1148,44 @@ def test_no_mem_leak_with_double_prefetch(self, close: bool):
       if close:
         it.close()  # pytype: disable=attribute-error
 
+  @parameterized.parameters(["naive", "close", "context"])
+  def test_early_break_continues_prefetching(self, method: str):
+    count = 0
+
+    class SlowCountingSource:
+      def __len__(self):
+        return 16
+
+      def __getitem__(self, index):
+        nonlocal count
+        time.sleep(0.1)
```

Comment:
sleep releases GIL, so this will run with actual multithreading and the thread-unsafe update below may make this test flaky. a bit more stable test could be to get a refcount before and after close for the source object

Reply:
I've added a `threading.Lock` around mutation of the `count` object.